dali/imgcodec/decoders/nvjpeg.cc

```diff
@@ -0,0 +1,169 @@
+// Copyright (c) 2019-2022, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "dali/imgcodec/decoders/nvjpeg.h"
+#include "dali/imgcodec/decoders/nvjpeg/nvjpeg_helper.h"
+#include "dali/imgcodec/decoders/nvjpeg/nvjpeg_memory.h"
+#include "dali/imgcodec/decoders/nvjpeg/permute_layout.h"
+
+namespace dali {
+namespace imgcodec {
+
+NvJpegDecoderInstance::NvJpegDecoderInstance(int device_id, ThreadPool *tp)
+: BatchParallelDecoderImpl(device_id, tp),
+  device_allocator_(nvjpeg_memory::GetDeviceAllocator()),
+  pinned_allocator_(nvjpeg_memory::GetPinnedAllocator()),
+  resources_(tp->NumThreads()) {
+  CUDA_CALL(nvjpegCreateSimple(&nvjpeg_handle_));
+
+  for (auto &resource : resources_) {
+    CUDA_CALL(nvjpegJpegStreamCreate(nvjpeg_handle_, &resource.jpeg_stream));
+    CUDA_CALL(nvjpegBufferPinnedCreate(nvjpeg_handle_, &pinned_allocator_,
+                                       &resource.pinned_buffer));
+    CUDA_CALL(nvjpegBufferDeviceCreate(nvjpeg_handle_, &device_allocator_,
+                                       &resource.device_buffer));
+    CUDA_CALL(cudaStreamCreateWithFlags(&resource.stream, cudaStreamNonBlocking));
+
+    CUDA_CALL(cudaEventCreate(&resource.decode_event));
```

Comment:
If you follow the suggestion in the struct definition:
```suggestion
    resource.decode_event = CudaEvent::Create(device_id_);
```